2. Battery Sensor Data
The following readings were collected during a 6-hour test:
Time (h) Temperature (°C) Voltage (V) Current (A) Charge (%)
0 32 400 20 95
1 35 395 22 88
2 38 390 25 80
3 40 385 28 70
4 42 380 30 60
5 45 375 32 48

Task 1 — Analysis Using Python Basics
Store the given sensor readings using appropriate Python data structures, then perform the following tasks.
------------------------------------------------
1. Display the sensor readings
Use a for loop to display all battery readings.


In [8]:
import pandas as pd

data = {
    "Time (h)": [0, 1, 2, 3, 4, 5],
    "Temperature (°C)": [32, 35, 38, 40, 42, 45],
    "Voltage (V)": [400, 395, 390, 385, 380, 375],
    "Current (A)": [20, 22, 25, 28, 30, 32],
    "Charge (%)": [95, 88, 80, 70, 60, 48]
}

df = pd.DataFrame(data)
df

,Time (h),Temperature (°C),Voltage (V),Current (A),Charge (%)
0,0,32,400,20,95
1,1,35,395,22,88
2,2,38,390,25,80
3,3,40,385,28,70
4,4,42,380,30,60
5,5,45,375,32,48


In [9]:
for index, row in df.iterrows():
    print("Time:", row["Time (h)"], "h",
          "Temperature:", row["Temperature (°C)"], "°C",
          "Voltage:", row["Voltage (V)"], "V",
          "Current:", row["Current (A)"], "A",
          "Charge:", row["Charge (%)"], "%")

Time: 0 h Temperature: 32 °C Voltage: 400 V Current: 20 A Charge: 95 %
Time: 1 h Temperature: 35 °C Voltage: 395 V Current: 22 A Charge: 88 %
Time: 2 h Temperature: 38 °C Voltage: 390 V Current: 25 A Charge: 80 %
Time: 3 h Temperature: 40 °C Voltage: 385 V Current: 28 A Charge: 70 %
Time: 4 h Temperature: 42 °C Voltage: 380 V Current: 30 A Charge: 60 %
Time: 5 h Temperature: 45 °C Voltage: 375 V Current: 32 A Charge: 48 %


2. Identify high temperature
Apply the following condition:

If T &gt; 40 °C

Display: “High Temperature Alert”
Otherwise, display: “Temperature Normal”

In [15]:
for T in df["Temperature (°C)"]:
    if T > 40:
        print(T,"High Temperature Alert")
    else:
        print(T,"Temperature Normal")

32 Temperature Normal
35 Temperature Normal
38 Temperature Normal
40 Temperature Normal
42 High Temperature Alert
45 High Temperature Alert


3. Identify low battery
Apply the following condition:

If C &lt; 50 %

Display: “Low Battery Alert”
Otherwise, display: “Battery Charge Normal”

In [14]:
for C in df["Charge (%)"]:
    if C < 50:
        print(C,"Low Battery Alert")
    else:
        print(C,"Battery Charge Normal")

95 Battery Charge Normal
88 Battery Charge Normal
80 Battery Charge Normal
70 Battery Charge Normal
60 Battery Charge Normal
48 Low Battery Alert


4. Create a function
Create a function battery_status(charge) that returns a status based on:
High, C &gt; 70
Status = Medium, 50 ≤ C ≤ 70
Low, C &lt; 50

In [18]:
def battery_status(charge):
     if charge > 70:
         return "High"
     elif 50 <= charge <= 70:
         return "Medium"
     else:
         return "Low"

for C in df["Charge (%)"]:
     print("Charge:", C, "% - Status:", battery_status(C))

Charge: 95 % - Status: High
Charge: 88 % - Status: High
Charge: 80 % - Status: High
Charge: 70 % - Status: Medium
Charge: 60 % - Status: Medium
Charge: 48 % - Status: Low


Task 2 — Analysis Using NumPy
------------------------------------------------
Convert the battery data into a NumPy array. Assume the data matrix is:
┌ 32 400 20 95 ┐
│ 35 395 22 88 │
D = │ 38 390 25 80 │
│ 40 385 28 70 │
│ 42 380 30 60 │
└ 45 375 32 48 ┘
where the columns represent: [ Temperature, Voltage, Current, Charge ]
Perform the following operations:
1. Find the shape of the array
Determine the number of rows and the number of columns.

In [23]:
import numpy as np

D = np.array([
     [32, 400, 20, 95],
     [35, 395, 22, 88],
     [38, 390, 25, 80],
     [40, 385, 28, 70],
     [42, 380, 30, 60],
     [45, 375, 32, 48]
     ])
print("Shape of the array:", D.shape)
rows, columns = D.shape
print("Number of rows:", rows)
print("Number of columns:", columns)

Shape of the array: (6, 4)
Number of rows: 6
Number of columns: 4


2. Calculate the average of each parameter
For example, the average temperature is:

T̄ = (T₁ + T₂ + T₃ + T₄ + T₅ + T₆) / 6
Similarly, calculate V̄ (average voltage), Ī (average current), and C̄ (average charge).

In [24]:
average_temperature = np.mean(D[:, 0])
average_voltage = np.mean(D[:, 1])
average_current = np.mean(D[:, 2])
average_charge = np.mean(D[:, 3])

print("Average Temperature:", average_temperature, "°C")
print("Average Voltage:", average_voltage, "V")
print("Average Current:", average_current, "A")
print("Average Charge:", average_charge, "%")

Average Temperature: 38.666666666666664 °C
Average Voltage: 387.5 V
Average Current: 26.166666666666668 A
Average Charge: 73.5 %


3. Find the maximum and minimum temperature
Determine T_max and T_min.

In [25]:
T_max = np.max(D[:, 0])
T_min = np.min(D[:, 0])

print("Maximum Temperature:", T_max, "°C")
print("Minimum Temperature:", T_min, "°C")

Maximum Temperature: 45 °C
Minimum Temperature: 32 °C


4. Extract the voltage column
Extract the column vector:

┌ 400 ┐
│ 395 │
V = │ 390 │
│ 385 │
│ 380 │
└ 375 ┘

In [26]:
for voltage in D[:, 1]:
    print(voltage, "V")

400 V
395 V
390 V
385 V
380 V
375 V


5. Identify high-temperature readings
Find all rows satisfying:

T &gt; 40 °C

In [28]:
high_temperature = D[D[:, 0] > 40]
print(high_temperature)

[[ 42 380  30  60]
 [ 45 375  32  48]]


6. Identify low battery readings
Find all rows satisfying:

C &lt; 50 %

In [30]:
low_battery_reading = D[D[:, 3] < 50]
print(low_battery_reading)

[[ 45 375  32  48]]


Task 3 — Analysis Using Linear Algebra
The company uses temperature and current to calculate a simplified Battery Stress Index.
------------------------------------------------
Let the transformation matrix be:

┌ 2 1 ┐
A = │ │
└ 1 3 ┘

Consider the sensor input:

┌ T ┐ ┌ 40 ┐
X = │ │ = │ │
└ I ┘ └ 25 ┘

The output is calculated as:

Y = A X

Therefore:

┌ 2 1 ┐ ┌ 40 ┐
Y = │ │ × │ │
└ 1 3 ┘ └ 25 ┘

Students must calculate the resulting vector Y.
Additional Linear Algebra Tasks
For the matrix A above, find:
1. Transpose
Aᵀ

In [31]:
A = np.array([[2, 1], [1, 3]])
A_transpose = A.T
print("Transpose of A:")
print(A_transpose)

Transpose of A:
[[2 1]
 [1 3]]


2. Determinant

det(A)

In [35]:
determinant = np.linalg.det(A)
print("Determinant of A:", round(determinant))

Determinant of A: 5


3. Inverse
A⁻¹, if the inverse exists.

In [38]:
inverse_of_A = np.linalg.inv(A)
print("Inverse of A:")
print(np.array(inverse_of_A))

Inverse of A:
[[ 0.6 -0.2]
 [-0.2  0.4]]


Solving a System of Equations
The engineers need to determine two unknown battery parameters x and y. The equations are:

2x + y = 100
x + 3y = 120

Write the equations in matrix form:

A X = B

where:

┌ 2 1 ┐ ┌ x ┐ ┌ 100 ┐
A = │ │ X = │ │ B = │ │
└ 1 3 ┘ └ y ┘ └ 120 ┘

Students must solve for x and y using NumPy.

In [39]:
import numpy as np

A = np.array([ [2, 1], [1, 3] ])
B = np.array([100, 120])

X = np.linalg.solve(A, B)

print("Value of x:", X[0])
print("Value of y:", X[1])

Value of x: 36.0
Value of y: 28.0


Task 4 — Analysis Using Calculus
The battery charge is modeled as a function of time:
------------------------------------------------

C(t) = 100 − 2t − ½t²
where C(t) is the battery charge percentage and t is the time in hours.
1. Find the rate of change of battery charge
The rate at which the battery charge changes is given by dC/dt. Differentiate C(t) and determine dC/dt.

In [40]:
import sympy as sp

t = sp.symbols('t')

C = 100 - 2*t - (1/2)*t**2

dC_dt = sp.diff(C, t)

print("Battery Charge Function:", C)
print("Rate of Change (dC/dt):", dC_dt)

Battery Charge Function: -0.5*t**2 - 2*t + 100
Rate of Change (dC/dt): -1.0*t - 2


2. Calculate the discharge rate
Find the discharge rate at t = 2 and t = 5, that is, calculate:

dC/dt at t = 2 and dC/dt at t = 5

3. Interpret the result
Based on the values of dC/dt, determine whether the battery is losing charge:
● At a constant rate
● At an increasing rate
● At a decreasing rate

In [43]:
rate_t2 = dC_dt.subs(t, 2)
rate_t5 = dC_dt.subs(t, 5)
print("Rate of change at t = 2:", round(rate_t2), "% per hour")
print("Rate of change at t = 5:", round(rate_t5), "% per hour")
#Increasing rate

Rate of change at t = 2: -4 % per hour
Rate of change at t = 5: -7 % per hour


Final Analysis and Decision
------------------------------------------------

After completing all four tasks, prepare a Battery Performance Report. Answer the following questions:
1. Safety
At what time does the temperature first exceed 40 °C?

ANS:- 4hrs


------------------------------------------------

2. Battery Charge
Calculate the total charge reduction:

Charge Loss = C_initial − C_final

ANS:- Charge loss = 95-48 = 47%



------------------------------------------------

3. NumPy Analysis
Calculate T̄, V̄, Ī, and C̄.

ANS:-
Average Temperature: 38.666666666666664 °C

Average Voltage: 387.5 V

Average Current: 26.166666666666668 A

Average Charge: 73.5 %


------------------------------------------------

4. Linear Algebra
Calculate the transformed output Y = AX, and solve AX = B.

ANS:-
Solution X:
[36. 28.]

Transformed output Y = AX:
[100. 120.]




------------------------------------------------
5. Calculus
Analyze the rate of battery discharge using dC/dt.

ANS:-
The discharge rate changes from −4%/hour to −7%/hour, meaning the battery is losing charge at an increasing rate.


------------------------------------------------
6. Final Decision
Based on the analysis, select one of the following:
● APPROVE — The battery is operating safely.
● MONITOR — The battery requires continuous monitoring.
● REJECT — The battery shows unsafe operating conditions.

ANS:-Monitor